### PR TITLE
add async support to api

### PR DESCRIPTION
**Pull request by rodnaph**

### CODE REVIEW

File: src/GenericApi.php [hunk 1]
```diff
@@ -110,38 +111,90 @@ private function getUri(array $uriParameters = [])
...
         $queryParameters['limit'] = isset($queryParameters['limit']) ? $queryParameters['limit'] : 100;
         $paginator = $this->createPaginator($queryParameters, $uriParameters);
-        $results = $paginator->getCurrentPageResults();
-        while ($paginator->hasNextPage()) {
-            $paginator->nextPage();
-            $results = array_merge($results, $paginator->getCurrentPageResults());
-        }
 
-        return $results;
+        $promise = new Promise(function() use ($paginator, $concurrency, &$promise) {
```
Comment (bendavies) [inline src/GenericApi.php]: can just be returned.

Comment (rodnaph) [inline src/GenericApi.php]: no, i need a reference to the variable, it's used in the wait callback to resolve the promise.

File: src/GenericApi.php [hunk 2]
```diff
@@ -110,38 +111,90 @@ private function getUri(array $uriParameters = [])
...
+                    $paginator->nextPage();
+
+                    yield $paginator->getCurrentPageResultsAsync();
+                }
+            })();
+
+            (new EachPromise($promises, [
+                'concurrency' => $concurrency,
+                'fulfilled' => function ($response) use (&$result) {
+                    $result = \array_merge($result, $response);
```
Comment (bendavies) [inline src/GenericApi.php]: this is actually very very slow on trivially large results sets. better is:

`$result[] = $response;`
then `promise->resolve(call_user_func_array('array_merge', $result))`

File: src/ApiInterface.php [hunk 3]
```diff
@@ -34,13 +44,27 @@ public function getAll(array $queryParameters = [], array $uriParameters = []);
     /**
      * @param  array $queryParameters
      * @param  array $uriParameters
+     * @return Promise
+     */
+    public function getAllAsync(array $queryParameters = [], array $uriParameters = [], $concurrency = 1);
```
Comment (bendavies) [inline src/ApiInterface.php]: missing doc block for concurrency

Comment (bendavies) [inline src/ApiInterface.php]: type hint it

File: src/GenericApi.php [hunk 4]
```diff
@@ -110,38 +111,90 @@ private function getUri(array $uriParameters = [])
...
      */
     public function getAll(array $queryParameters = [], array $uriParameters = [])
     {
+        return $this->getAllAsync($queryParameters, $uriParameters)->wait();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAllAsync(array $queryParameters = [], array $uriParameters = [], $concurrency = 1)
```
Comment (bendavies) [inline src/GenericApi.php]: type hint it

File: src/GenericApi.php (hunk 2 above)
Comment (rodnaph) [inline src/GenericApi.php]: Holy shit you're right, just tried this...

```
<?php

$result = [];
$data = array_fill(0, 1000, 'X');

for ($i=0; $i<1000; $i++) {
    $result = array_merge($result, $data);
}

// 11 seconds

for ($i=0; $i<1000; $i++) {
    $result[] = $data;
}

call_user_func_array('array_merge', $result);

// 0.063 seconds
```

Wowzer!  Use `array_merge` sparingly then...

File: src/ApiInterface.php (hunk 3 above)
Comment (rodnaph) [inline src/ApiInterface.php]: done

File: src/GenericApi.php (hunk 2 above)
Comment (bendavies) [inline src/GenericApi.php]: hehe, did you doubt me?
Just be careful using it in a loop, basically.

**Pull request merged**